### DISCUSSION TITLE
Is there a way to pass custom Field properties to ruleProcessor?

### DISCUSSION THREAD

**Discussion by mateioprea** [react-querybuilder/react-querybuilder#444] [Q&A] [answered]
Hi!

For example, I have the following fields configuration: 

```
{
    name: 'firstName',
    label: 'First Name',
    placeholder: 'Enter first name',
    validator,
    fieldData: {
        type: 'string',
        anotherProp: 'another one'
    }
  }
  ```

```
transformQuery(query, {
    ruleProcessor: r => {
      console.log(r)
    },
  })
```

logs to the console: 

```
{
    "id": "bd3f816d-1308-4896-9b56-d67330aa2025",
    "field": "firstName",
    "operator": "=",
    "valueSource": "value",
    "value": "",
    "path": [
        11
    ]
}
```

But i want this to also include "fieldData" property as well.

Thank you! 

Later edit: 

After investigating a bit, I found that the createRule does not allow passing of arbitrary properties into rules and it forces the needed properties. Is there a way to bypass this?

**Comment by mateioprea** [comment]
To answer my own question, one way around this will be to extract the needed properties in the ```ruleProcessor``` method by looking at the field definitions array.

**Comment by jakeboone02** [comment, marked as answer]
You can use the [`onAddRule` prop](https://react-querybuilder.js.org/docs/api/querybuilder#onaddrule) to manipulate the rule object that gets added (and `onAddGroup` for groups). If you want to add arbitrary properties to the rule, do something like this:

```jsx
const onAddRule = (rule) => ({
  ...rule,
  // I think you alluded to this in your own answer:
  fieldData: fields.find(f => f.name === rule.field).fieldData,
});

const App = () => (
  <QueryBuilder fields={fields} onAddRule={onAddRule} />
);
```

Just note that the `fieldData` property of the new rule will never by updated automatically by the component. If the user updates the field, the rule's `fieldData` may be outdated.

**Reply by mateioprea** [reply]
Thank you!
